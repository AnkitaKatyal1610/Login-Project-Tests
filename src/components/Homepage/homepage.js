import React, { Component } from 'react'
import { Button, Table, Image, Pagination, Icon } from 'semantic-ui-react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { withRouter } from 'react-router-dom'

import { logoutAction } from '../../action/action'
import { getUsersAction } from '../../action/userActions'

export class Homepage extends Component {
    state = {
        page: 1
    }

    componentWillMount() {
        let token = localStorage.getItem('token');
        if (this.props.token === '' && !token) {
            this.props.history.replace('/login');
        }
        if (this.props.getUsersAction)
            this.props.getUsersAction(this.state.page);

    }
    logoutClicked = () => {
        if (this.props.logoutAction) {
            this.props.logoutAction();
            this.props.history.replace('/login');
        }
    }

    setPageNum = (e, data) => {
        this.setState({ page: data.activePage });
        this.props.getUsersAction(data.activePage);
    }

    render() {
        return (
            <div>
                <h2>Welcome, you have logged in successfully</h2>{" "}
                <Button onClick={this.logoutClicked} color='orange'>Logout</Button>
                <Table textAlign='center' inverted>
                    <Table.Header>
                        <Table.Row>
                            <Table.HeaderCell>#</Table.HeaderCell>
                            <Table.HeaderCell>Avatar</Table.HeaderCell>
                            <Table.HeaderCell>First Name</Table.HeaderCell>
                            <Table.HeaderCell>Last Name</Table.HeaderCell>
                            <Table.HeaderCell>Actions</Table.HeaderCell>
                        </Table.Row>
                    </Table.Header>
                    <Table.Body>
                        {
                            (!this.props.users) ? null : this.props.users.map((user, i) => {
                                return (
                                    <Table.Row key={i}>
                                        <Table.Cell>{i + 1 + (this.state.page - 1) * 3}</Table.Cell>
                                        <Table.Cell><Image centered src={user.avatar} size='mini' circular /></Table.Cell>
                                        <Table.Cell>{user.first_name}</Table.Cell>
                                        <Table.Cell>{user.last_name}</Table.Cell>
                                        <Table.Cell><Button color='blue'>Edit</Button><Button color='red'>Delete</Button></Table.Cell>
                                    </Table.Row>
                                )
                            })
                        }
                    </Table.Body>
                </Table>
                <Pagination defaultActivePage={this.state.page} onPageChange={this.setPageNum}
                    totalPages={4} prevItem={{ content: <Icon name="angle left" size='small' />, icon: true }}
                    nextItem={{ content: <Icon name="angle right" size='small' />, icon: true }}
                    firstItem={{ content: <Icon name="caret left" size='small' />, icon: true }}
                    lastItem={{ content: <Icon name="caret right" size='small' />, icon: true }} />
            </div>
        );
    }
}

const mapState = (state) => {
    return {
        token: state.auth.token,
        users: state.user.users
    }
}
const mapDispatch = (dispatch) => {
    return {
        logoutAction: bindActionCreators(logoutAction, dispatch),
        getUsersAction: bindActionCreators(getUsersAction, dispatch)
    }
}
export default withRouter(connect(mapState, mapDispatch)(Homepage))