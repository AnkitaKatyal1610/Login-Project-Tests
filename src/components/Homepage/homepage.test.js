import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

import { Homepage } from './homepage';

Enzyme.configure({ adapter: new Adapter() })
let props = {
    getUsersAction: jest.fn(),
    token: "abcd",
    users: [],
    logoutAction: jest.fn(),
    history: {
        replace: jest.fn()
    }
}

let wrapper;
beforeEach(() => {
    wrapper = shallow(<Homepage {...props} />)
})

describe('Testing Component Homepage', () => {
    it('matches snapshot', () => {
        const tree = renderer.create(<Homepage {...props} />).toJSON();
        expect(tree).toMatchSnapshot();

    })
    it('has welcome text for the logged in user', () => {
        let welcomeText = wrapper.find('h2').text();
        expect(welcomeText).toBe("Welcome, you have logged in successfully");
    })

    it('has a logout button', () => {
        let button = wrapper.find('Button');
        expect(button.length).toEqual(1);
        expect(button.prop('children')).toBe('Logout');
    })

    it('calls action to display users when the component mounts', () => {
        expect(props.getUsersAction).toHaveBeenCalled();
    })
    it('displays a table with user information', () => {
        let tables = wrapper.find('Table');
        expect(tables.length).toEqual(1)
    })

    it('displays a pagination component', () => {
        let pages = wrapper.find('Pagination');
        expect(pages.length).toEqual(1);
    })

    it('changes page number in state when page button is clicked and calls action to get users of that page', () => {
        let pages = wrapper.find('Pagination');
        pages.prop('onPageChange', { data: { activePage: 3 } })
        let page = pages.prop('activePage');
        expect(wrapper.state().page).toEqual(page);
        expect(props.getUsersAction).toHaveBeenCalled();
    })

    it('calls onClick handler when logout button is clicked', () => {
        let spy = jest.spyOn(wrapper.instance(), 'logoutClicked')
        wrapper.update();
        wrapper.instance().forceUpdate();
        let button = wrapper.find('Button');
        button.simulate('click');
        expect(spy).toHaveBeenCalled();
    })

    it('calls logout action when logout button is clicked', () => {
        let button = wrapper.find('Button');
        button.simulate('click');
        expect(props.logoutAction).toHaveBeenCalled();
    })
})