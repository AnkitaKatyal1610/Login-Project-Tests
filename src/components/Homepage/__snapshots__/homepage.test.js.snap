// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Component Homepage matches snapshot 1`] = `
<div
  style={
    Object {
      "margin": "15px",
    }
  }
>
  <h2>
    Welcome, you have logged in successfully
  </h2>
   
  <button
    className="ui orange button"
    onClick={[Function]}
  >
    Logout
  </button>
  <table
    className="ui inverted center aligned table"
  >
    <thead
      className=""
    >
      <tr
        className=""
      >
        <th
          className=""
        >
          #
        </th>
        <th
          className=""
        >
          Avatar
        </th>
        <th
          className=""
        >
          First Name
        </th>
        <th
          className=""
        >
          Last Name
        </th>
        <th
          className=""
        >
          Actions
        </th>
      </tr>
    </thead>
    <tbody
      className=""
    />
  </table>
  <div
    aria-label="Pagination Navigation"
    className="ui pagination menu"
    role="navigation"
  >
    <a
      aria-current={false}
      aria-disabled={false}
      className="icon item"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="firstItem"
      value={1}
    >
      <i
        aria-hidden="true"
        className="caret left small icon"
        onClick={[Function]}
      />
    </a>
    <a
      aria-current={false}
      aria-disabled={false}
      className="icon item"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="prevItem"
      value={1}
    >
      <i
        aria-hidden="true"
        className="angle left small icon"
        onClick={[Function]}
      />
    </a>
    <a
      aria-current={true}
      aria-disabled={false}
      className="active item"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="pageItem"
      value={1}
    >
      1
    </a>
    <a
      aria-current={false}
      aria-disabled={false}
      className="item"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="pageItem"
      value={2}
    >
      2
    </a>
    <a
      aria-current={false}
      aria-disabled={false}
      className="item"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="pageItem"
      value={3}
    >
      3
    </a>
    <a
      aria-current={false}
      aria-disabled={false}
      className="item"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="pageItem"
      value={4}
    >
      4
    </a>
    <a
      aria-current={false}
      aria-disabled={false}
      className="icon item"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="nextItem"
      value={2}
    >
      <i
        aria-hidden="true"
        className="angle right small icon"
        onClick={[Function]}
      />
    </a>
    <a
      aria-current={false}
      aria-disabled={false}
      className="icon item"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="lastItem"
      value={4}
    >
      <i
        aria-hidden="true"
        className="caret right small icon"
        onClick={[Function]}
      />
    </a>
  </div>
</div>
`;
